File: src/platform.ts
import { invoke } from "@tauri-apps/api/core";
import { useEffect, useState } from "react";

export interface PlatformInfo {
  os: "macos" | "windows" | "linux" | "unknown";
  session: "native" | "wayland" | "x11" | "unknown";
  trayBackend: "native" | "legacy-gtk" | "appindicator";
  supportsTrayClickActions: boolean;
  supportsNativePopupCorners: boolean;
}

const fallbackPlatform: PlatformInfo = {
  os: "unknown",
  session: "unknown",
  trayBackend: "native",
  supportsTrayClickActions: false,
  supportsNativePopupCorners: false,
};

let cachedPlatform: PlatformInfo | null = null;
let platformRequest: Promise<PlatformInfo> | null = null;

export function getPlatformInfo(): Promise<PlatformInfo> {
  if (cachedPlatform) return Promise.resolve(cachedPlatform);
  if (!platformRequest) {
    platformRequest = invoke<PlatformInfo>("get_platform_info")
      .then((platform) => {
        cachedPlatform = platform;
        return platform;
      })
      .catch(() => {
        cachedPlatform = fallbackPlatform;
        return fallbackPlatform;
      });
  }
  return platformRequest;
}

export function currentPlatform(): PlatformInfo {
  return cachedPlatform ?? fallbackPlatform;
}

export function usePlatform(): PlatformInfo {
  const [platform, setPlatform] = useState(currentPlatform);
  useEffect(() => {
    void getPlatformInfo().then(setPlatform);
  }, []);
  return platform;
}

File: src-tauri/src/platform.rs
use serde::Serialize;

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlatformInfo {
    os: &'static str,
    session: &'static str,
    tray_backend: &'static str,
    supports_tray_click_actions: bool,
    supports_native_popup_corners: bool,
}

#[cfg(target_os = "linux")]
fn linux_session() -> &'static str {
    if let Ok(session) = std::env::var("XDG_SESSION_TYPE") {
        if session.eq_ignore_ascii_case("wayland") {
            return "wayland";
        }
        if session.eq_ignore_ascii_case("x11") {
            return "x11";
        }
    }
    if std::env::var("WAYLAND_DISPLAY").is_ok_and(|value| !value.is_empty()) {
        return "wayland";
    }
    if std::env::var("DISPLAY").is_ok_and(|value| !value.is_empty()) {
        return "x11";
    }
    "unknown"
}

#[tauri::command]
pub fn get_platform_info() -> PlatformInfo {
    #[cfg(target_os = "linux")]
    {
        let tray_backend = crate::tray::platform_tray_backend();
        PlatformInfo {
            os: "linux",
            session: linux_session(),
            tray_backend,
            supports_tray_click_actions: tray_backend == "legacy-gtk",
            supports_native_popup_corners: false,
        }
    }

    #[cfg(target_os = "macos")]
    {
        PlatformInfo {
            os: "macos",
            session: "native",
            tray_backend: "native",
            supports_tray_click_actions: true,
            supports_native_popup_corners: true,
        }
    }

    #[cfg(target_os = "windows")]
    {
        PlatformInfo {
            os: "windows",
            session: "native",
            tray_backend: "native",
            supports_tray_click_actions: true,
            supports_native_popup_corners: false,
        }
    }

    #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows")))]
    {
        PlatformInfo {
            os: "unknown",
            session: "unknown",
            tray_backend: "native",
            supports_tray_click_actions: false,
            supports_native_popup_corners: false,
        }
    }
}
